searx/network/context.py

```diff
@@ -0,0 +1,404 @@
+from typing import Callable, Optional, final
+
+try:
+    from typing import ParamSpec, TypeVar
+except ImportError:
+    # to support Python < 3.10
+    from typing_extensions import ParamSpec, TypeVar
+
+import httpx
+
+from searx.network.client import ABCHTTPClient, SoftRetryHTTPException
+
+P = ParamSpec('P')
+R = TypeVar('R')
+HTTPCLIENTFACTORY = Callable[[], ABCHTTPClient]
+
+DEFAULT_TIMEOUT = 120.0
+
+
+## NetworkContext
+
+
+class NetworkContext(ABC):
+    """Abstract implementation: the call must defined in concrete classes.
+
+    Lifetime: one engine request or initialization of an engine.
+    """
+
+    __slots__ = ('_retries', '_http_client_factory', '_http_client', '_start_time', '_http_time', '_timeout')
+
+    def __init__(
+        self,
+        retries: int,
+        http_client_factory: HTTPCLIENTFACTORY,
+        start_time: Optional[float],
+        timeout: Optional[float],
+    ):
+        self._retries: int = retries
+        # wrap http_client_factory here, so we can forget about this wrapping
+        self._http_client_factory = _TimeHTTPClientWrapper.wrap_factory(http_client_factory, self)
+        self._http_client: Optional[ABCHTTPClient] = None
+        self._start_time: float = start_time or default_timer()
+        self._http_time: float = 0.0
+        self._timeout: Optional[float] = timeout
+
+    @abstractmethod
+    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
+        """Call func within the network context.
+
+        The retry policy might call func multiple times.
+
+        Within the function self.get_http_client() returns an HTTP client to use.
+
+        The retry policy might send multiple times the same HTTP request
+        until it works or the retry count falls to zero.
+        """
+
+    @final
+    def request(self, *args, **kwargs):
+        """Convenient method to wrap a call to request inside the call method.
+
+        Use a new HTTP client to wrap a call to the request method using self.call
+        """
+        self._set_http_client()
+        return self.call(self._get_http_client().request, *args, **kwargs)
+
+    @final
+    def stream(self, *args, **kwargs):
+        """Convenient method to wrap a call to stream inside the call method.
+
+        Use a new HTTP client to wrap a call to the stream method using self.call
+        """
+        self._set_http_client()
+        return self.call(self._get_http_client().stream, *args, **kwargs)
+
+    @final
+    def get_http_runtime(self) -> Optional[float]:
+        """Return the amount of time spent on HTTP requests"""
+        return self._http_time
+
+    @final
+    def get_remaining_time(self, _override_timeout: Optional[float] = None) -> float:
+        """Return the remaining time for the context.
+
+        _override_timeout is not intended to be used outside this module.
+        """
+        timeout = _override_timeout or self._timeout or DEFAULT_TIMEOUT
+        timeout += 0.2  # overhead
```

Comment:
Do we have a use case / a need for `_override_timeout`? (I haven't found a usage of this argument).  May its better to have an argument `offset=0.2`?

Reply:
https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/context.py#L228-L231

called by:

https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/context.py#L211-L219

In English: the allows the user to override the timeout per HTTP request. For example 
```python
with searx.network.networkcontext_manager(timeout=2):
  searx.network.get("https://example.com", timeout=1) # the user ask for a timeout of 1 not 2.
```